Adds information on "script will never generate a response" errors (#15014)
cloudflare/cloudflare-docs: content/workers/observability/errors.md

A Worker cannot call itself or another Worker more than 16 times. In  order to prevent infinite loops between Workers, the [`CF-EW-Via`](/fundamentals/reference/http-request-headers/#cf-ew-via) header's value is an integer that indicates how many invocations are left. Every time a Worker is invoked, the integer will decrement by 1. If the count reaches zero, a [`1019`](#error-pages-generated-by-workers) error is returned.

### "The script will never generate a response" errors

Some requests may return a 1101 error with `The script will never generate a response` in the error message. This occurs when the Workers runtime detects that all the code associated with the request has executed and no events are left in the event loop, but a Response has not been returned.

#### Cause 1: Unresolved Promises

> This is most commonly caused by relying on a Promise that is never resolved or rejected, in order to return a Response. To debug, look for Promises within your code or dependencies' code that block a Response, and ensure they are resolved or rejected.

Reviewer: ```suggestion
This is most commonly caused by relying on a Promise that is never resolved or rejected, which is required to return a Response. To debug, look for Promises within your code or dependencies' code that block a Response, and ensure they are resolved or rejected.
```

```js
---
filename: index.js
highlight: [9]
---
export default {
  fetch(req) {
    let response = new Response("Example response");
    let { promise, resolve } = Promise.withResolvers();

    // If the promise is not resolved, the Workers runtime will
    // recognize this and throw an error.

    // setTimeout(resolve, 0)

    return promise.then(() => response);
  },
};
```

#### Cause 2: WebSocket connections that are never closed

If a WebSocket is missing the proper code to close its server-side connection, the Workers runtime will throw a `script will never generate a response` error. In the example below, the `'close'` event from the client is not properly handled by calling `server.close()`, and the error is thrown. In order to avoid this, ensure that the WebSocket's server-side connection is properly closed via an event listener or other server-side logic.

```js
---
filename: index.js
highlight: [10]
---
async function handleRequest(request) {
  let webSocketPair = new WebSocketPair();
  let [client, server] = Object.values(webSocketPair);
  server.accept();

>   server.addEventListener('close', () => {

Reviewer 2: Is this correct? The paragraph above implies this should be `client.addEventListener`.

Maybe this section needs a once over as it's not very easy to understand cc @irvinebroque

Reviewer 3: It doesn't work. I've tried this code and the error still exists.
```js
server.addEventListener("close", () => {
  console.log("This function will be run, but the error still exists.");
  server.close();
  client.close();
});
client.addEventListener("close", () => {
  console.log("This function will not be run.");
  server.close();
  client.close();
});
```

Reviewer 3: I have tried the code but the error still exists. It seems that there are no ways to catch the error when using WebSocket. Here is my code for reproduction.

Worker:
```ts
export default {
    async fetch(request, env, ctx): Promise<Response> {
        if (request.headers.get("Upgrade") !== "websocket")
            return new Response("Expected Upgrade: websocket", { status: 426 });
        const { 0: client, 1: server } = new WebSocketPair();
        server.accept();
        server.addEventListener("close", event => {
            console.log("Close event triggered");
            // All lines below are useless
            server.close();
            client.close();
            event.preventDefault();
            event.stopImmediatePropagation();
        });
        server.addEventListener("error", event => console.log("Never triggered"));
        client.addEventListener("close", event => console.log("Never triggered"));
        client.addEventListener("error", event => console.log("Never triggered"));
        return new Response(null, { status: 101, webSocket: client, });
    },
} satisfies ExportedHandler<Env>;
```

Node.js Client:
```js
new WebSocket("ws://localhost:8787").onopen = () => process.exit();
```
P.S. I have only tested with `wrangler dev`.

A Worker cannot call itself or another Worker more than 16 times. In  order to prevent infinite loops between Workers, the [`CF-EW-Via`](/fundamentals/reference/http-request-headers/#cf-ew-via) header's value is an integer that indicates how many invocations are left. Every time a Worker is invoked, the integer will decrement by 1. If the count reaches zero, a [`1019`](#error-pages-generated-by-workers) error is returned.

### "The script will never generate a response" errors

> Some requests may return a 1101 error with `The script will never generate a response` in the error message. This occurs when the runtime detects that all the code associated with the request has executed and no events are left in the event loop, but a Response has not been returned. This is most often a Response relying on a Promise that never gets resolved or rejected. To debug this problem, look for the creation of Promises within your own code or library code that block a Response, then and ensure their resolution. In other runtimes, this could result in indefinitely hanging code, but Workers error immediately to help with debugging.

Reviewer 4: ```suggestion
Some requests may return a 1101 error with `The script will never generate a response` in the error message. This occurs when the Workers runtime detects that all the code associated with the request has executed and no events are left in the event loop, but a Response has not been returned.

#### Cause 1: Unresolved Promises

This is most commonly caused by relying on a Promise that is never resolved or rejected, in order to return a Response. To debug, look for Promises within your code or dependencies' code that block a Response, and ensure they are resolved or rejected.

In browsers and other JavaScript runtimes, equivalent code will hang indefinitely, leading to both bugs and memory leaks. The Workers runtime throws an explicit error to help you debug.
```

> In the (admittedly contrived) example below, the Response relies on a Promise resolution that never happens. Uncommenting the `resolve` callback solves the issue.

Author: Debated adding "(admittedly contrived)" or not. I think it gives the docs some flavor/life, but understand if that's not the tone we want.

Reviewer 4: ```suggestion
In the example below, the Response relies on a Promise resolution that never happens. Uncommenting the `resolve` callback solves the issue.
```

```js
export default {
	fetch(req) {
		let response = new Response("Example response");
		let { promise, resolve } = Promise.withResolvers();

> 		// If this a resolve will never happen, the Workers runtime will

Reviewer 4: ```suggestion
		// If the promise is not resolved, the Workers runtime will
```

		// recognize this and throw an error.

> 		// setTimeout(resolve, 0)

Reviewer 4: You can highlight a specific line in a code block, maybe worth doing here.

		return promise.then(() => response);
	},
};
```

> A subset of these cases are related to open WebSocket connections. If a WebSocket is missing the proper code to close it's server-side connection, a `script will never generate a response` error is thrown. In the example below, the `'close'` event from the client is not properly handled with a `server.close()`, and the error is thrown. In order to avoid this, ensure that the WebSocket's server-side connection is properly closed via an event listener or other server-side logic.

Reviewer 4: ```suggestion
#### Cause 2: WebSocket connections that are never closed

If a WebSocket is missing the proper code to close it's server-side connection, the Workers runtime will throw a `script will never generate a response` error. In the example below, the `'close'` event from the client is not properly handled by calling `server.close()`, and the error is thrown. In order to avoid this, ensure that the WebSocket's server-side connection is properly closed via an event listener or other server-side logic.
```

Reviewer 5: Also small typo fix: `it's` -> `its`
